File: packages/server/src/features/playlists/repository.ts
import Database from 'better-sqlite3';
import { randomUUID } from 'node:crypto';
import type { Playlist, PlaylistResolveMode, PlaylistVisibility, SmartPlaylistRules } from '@sonarly/shared';
import { compileSmartPlaylist } from '../smart-playlists/compiler.js';

export function generateShareToken(): string {
  return randomUUID();
}

interface DbPlaylist {
  id: string;
  name: string;
  description: string | null;
  owner_id: string;
  visibility: PlaylistVisibility;
  share_token: string | null;
  is_smart: number;
  rules_json: string | null;
  resolve_mode: string | null;
  created_at: string;
  updated_at: string;
}

function normalizeResolveMode(value: string | null | undefined): PlaylistResolveMode {
  return value === 'query' ? 'query' : 'tracks';
}

function parseRules(json: string | null): SmartPlaylistRules | undefined {
  if (!json) return undefined;
  try {
    return JSON.parse(json) as SmartPlaylistRules;
  } catch {
    return undefined;
  }
}

export function getPlaylistById(db: Database.Database, id: string): Playlist | undefined {
  const row = db.prepare('SELECT * FROM playlists WHERE id = ?').get(id) as DbPlaylist | undefined;
  if (!row) return undefined;
  const isSmart = row.is_smart === 1;
  return {
    id: row.id,
    name: row.name,
    description: row.description ?? undefined,
    ownerId: row.owner_id,
    visibility: row.visibility,
    shareToken: row.share_token ?? undefined,
    songIds: isSmart ? [] : fetchStaticSongIds(db, id),
    isSmart,
    rules: isSmart ? parseRules(row.rules_json) : undefined,
    resolveMode: normalizeResolveMode(row.resolve_mode),
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

function fetchStaticSongIds(db: Database.Database, playlistId: string): string[] {
  return db.prepare('SELECT song_id FROM playlist_songs WHERE playlist_id = ? ORDER BY position')
    .pluck().all(playlistId) as string[];
}

// Smart playlists resolve user-scoped rule fields (rating, loved, playcount,
// lastplayed) against the owner's data by default, so shared and public
// viewers all receive the same curated track list. 'query' mode re-resolves
// live against each viewer's own data instead.
function rulesUserId(playlist: Playlist, viewerUserId: string): string {
  return playlist.resolveMode === 'query' ? viewerUserId : playlist.ownerId;
}

export function resolvePlaylistSongIds(db: Database.Database, playlist: Playlist, userId: string): string[] {
  if (playlist.isSmart && playlist.rules) {
    const compiled = compileSmartPlaylist(db, playlist.rules, rulesUserId(playlist, userId));
    return db.prepare(compiled.sql).pluck().all(...compiled.params) as string[];
  }
  return playlist.songIds;
}

export function resolvePlaylistSongCount(db: Database.Database, playlist: Playlist, userId: string): number {
  if (playlist.isSmart && playlist.rules) {
    const compiled = compileSmartPlaylist(db, playlist.rules, rulesUserId(playlist, userId));
    const row = db.prepare(compiled.songCountSql).get(...compiled.songCountParams) as { count: number } | undefined;
    return row?.count ?? 0;
  }
  return playlist.songIds.length;
}

// Mirrors resolvePlaylistSongCount's sources so the list view's duration and
// songCount always describe the same membership.
export function resolvePlaylistSongDuration(db: Database.Database, playlist: Playlist, userId: string): number {
  if (playlist.isSmart && playlist.rules) {
    const compiled = compileSmartPlaylist(db, playlist.rules, rulesUserId(playlist, userId));
    const ids = db.prepare(compiled.sql).pluck().all(...compiled.params) as string[];
    return sumSongDurations(db, ids);
  }
  return sumSongDurations(db, playlist.songIds);
}

function sumSongDurations(db: Database.Database, songIds: string[]): number {
  if (songIds.length === 0) return 0;
  let total = 0;
  for (let i = 0; i < songIds.length; i += 500) {
    const chunk = songIds.slice(i, i + 500);
    const row = db.prepare(`
      SELECT COALESCE(SUM(duration), 0) AS total FROM songs WHERE id IN (${chunk.map(() => '?').join(',')})
    `).get(...chunk) as { total: number } | undefined;
    total += row?.total ?? 0;
  }
  return total;
}

export function createPlaylist(db: Database.Database, playlist: Playlist): void {
  const isSmart = playlist.isSmart === true;
  db.transaction(() => {
    db.prepare(`
      INSERT INTO playlists (id, name, description, owner_id, visibility, share_token, is_smart, rules_json, resolve_mode)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
    `).run(
      playlist.id,
      playlist.name,
      playlist.description ?? null,
      playlist.ownerId,
      playlist.visibility,
      playlist.shareToken ?? null,
      isSmart ? 1 : 0,
      isSmart && playlist.rules ? JSON.stringify(playlist.rules) : null,
      normalizeResolveMode(playlist.resolveMode),
    );
    if (!isSmart) {
      insertPlaylistSongs(db, playlist.id, playlist.songIds);
    }
  })();
}

export function updatePlaylist(db: Database.Database, playlist: Playlist): void {
  const isSmart = playlist.isSmart === true;
  db.transaction(() => {
    db.prepare(`
      UPDATE playlists
      SET name = ?, description = ?, visibility = ?, share_token = ?, is_smart = ?, rules_json = ?, resolve_mode = ?, updated_at = datetime('now')
      WHERE id = ?
    `).run(
      playlist.name,
      playlist.description ?? null,
      playlist.visibility,
      playlist.shareToken ?? null,
      isSmart ? 1 : 0,
      isSmart && playlist.rules ? JSON.stringify(playlist.rules) : null,
      normalizeResolveMode(playlist.resolveMode),
      playlist.id,
    );
    // Members only exist for standard playlists; clear stale rows when smart.
    db.prepare('DELETE FROM playlist_songs WHERE playlist_id = ?').run(playlist.id);
    if (!isSmart) {
      insertPlaylistSongs(db, playlist.id, playlist.songIds);
    }
  })();
}

function insertPlaylistSongs(db: Database.Database, playlistId: string, songIds: string[]): void {
  if (songIds.length === 0) return;
  const stmt = db.prepare('INSERT INTO playlist_songs (playlist_id, song_id, position) VALUES (?, ?, ?)');
  for (let i = 0; i < songIds.length; i++) {
    stmt.run(playlistId, songIds[i], i);
  }
}

// Smart playlists don't populate playlist_songs — their membership resolves
// from the stored rules at request time. Resolution results are cached briefly
// (keyed by playlist id + rules, so rule edits invalidate immediately) to avoid
// recompiling per track on cover-art-heavy pages.
const SMART_GRANT_CACHE_TTL_MS = 30_000;

interface SmartGrantCacheEntry {
  ids: Set<string>;
  expiresAt: number;
}

const smartGrantCache = new Map<string, SmartGrantCacheEntry>();

interface SmartLinkPlaylistRow {
  id: string;
  owner_id: string;
  rules_json: string | null;
}

function fetchSmartLinkPlaylists(db: Database.Database, shareToken: string): SmartLinkPlaylistRow[] {
  return db.prepare(`
    SELECT id, owner_id, rules_json
    FROM playlists
    WHERE visibility = 'link' AND share_token = ? AND is_smart = 1
  `).all(shareToken) as SmartLinkPlaylistRow[];
}

function resolveSmartGrantSongIds(db: Database.Database, row: SmartLinkPlaylistRow): Set<string> {
  const key = `${row.id}:${row.rules_json ?? ''}`;
  const now = Date.now();
  const cached = smartGrantCache.get(key);
  if (cached && cached.expiresAt > now) return cached.ids;

  const ids = new Set<string>();
  const rules = parseRules(row.rules_json);
  if (rules) {
    // Same compiler path as resolvePlaylistSongIds; user-scoped rule fields
    // resolve against the playlist owner, matching the anonymous detail view.
    const compiled = compileSmartPlaylist(db, rules, row.owner_id);
    for (const id of db.prepare(compiled.sql).pluck().all(...compiled.params) as string[]) {
      ids.add(id);
    }
  }
  smartGrantCache.set(key, { ids, expiresAt: now + SMART_GRANT_CACHE_TTL_MS });
  return ids;
}

function shareTokenGrantsSmartSongIds(db: Database.Database, shareToken: string): Set<string> | undefined {
  const smartPlaylists = fetchSmartLinkPlaylists(db, shareToken);
  if (smartPlaylists.length === 0) return undefined;
  const ids = new Set<string>();
  for (const row of smartPlaylists) {
    for (const id of resolveSmartGrantSongIds(db, row)) {
      ids.add(id);
    }
  }
  return ids;
}

// A share token only ever authorizes content belonging to the link-shared
// playlist it was minted for; both checks scope the EXISTS to that playlist.
export function shareTokenGrantsSong(db: Database.Database, shareToken: string, songId: string): boolean {
  const row = db.prepare(`
    SELECT 1
    FROM playlists p
    JOIN playlist_songs ps ON ps.playlist_id = p.id
    JOIN songs s ON s.id = ps.song_id AND s.active = 1
    WHERE p.visibility = 'link' AND p.share_token = ? AND ps.song_id = ?
    LIMIT 1
  `).get(shareToken, songId);
  if (row !== undefined) return true;
  return shareTokenGrantsSmartSongIds(db, shareToken)?.has(songId) ?? false;
}

export function shareTokenGrantsCoverArt(db: Database.Database, shareToken: string, coverArtId: string): boolean {
  const row = db.prepare(`
    SELECT 1
    FROM playlists p
    JOIN playlist_songs ps ON ps.playlist_id = p.id
    JOIN songs s ON s.id = ps.song_id AND s.active = 1
    LEFT JOIN albums a ON a.id = s.album_id
    WHERE p.visibility = 'link' AND p.share_token = ?
      AND (s.cover_art_id = ? OR a.cover_art_id = ?)
    LIMIT 1
  `).get(shareToken, coverArtId, coverArtId);
  if (row !== undefined) return true;

  const smartSongIds = shareTokenGrantsSmartSongIds(db, shareToken);
  if (!smartSongIds || smartSongIds.size === 0) return false;
  const ids = [...smartSongIds];
  for (let i = 0; i < ids.length; i += 500) {
    const chunk = ids.slice(i, i + 500);
    const match = db.prepare(`
      SELECT 1
      FROM songs s
      LEFT JOIN albums a ON a.id = s.album_id
      WHERE s.active = 1 AND s.id IN (${chunk.map(() => '?').join(',')})
        AND (s.cover_art_id = ? OR a.cover_art_id = ?)
      LIMIT 1
    `).get(...chunk, coverArtId, coverArtId);
    if (match !== undefined) return true;
  }
  return false;
}

export function sharePlaylistWithUser(db: Database.Database, playlistId: string, userId: string, canEdit: boolean): void {
  db.prepare(`
    INSERT INTO playlist_shares (playlist_id, user_id, can_edit) VALUES (?, ?, ?)
    ON CONFLICT(playlist_id, user_id) DO UPDATE SET can_edit = excluded.can_edit
  `).run(playlistId, userId, canEdit ? 1 : 0);
}

File: v2/internal/modules/playlists/repository.go
package playlists

import (
	"context"
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"

	"github.com/google/uuid"
	"github.com/miquelrosell99/sonarly/v2/internal/modules/auth"
)

// ErrNotFound is the single not-found sentinel: routes answer 404 both when
// a playlist is missing and when the caller has no access, so inaccessible
// ids cannot be probed (v2 contract: 404, not 403).
var ErrNotFound = errors.New("playlists: not found")

// ErrForbidden marks an authenticated caller who can view but not edit, or
// who attempts an owner-only operation on someone else's playlist (403).
var ErrForbidden = errors.New("playlists: forbidden")

// listRow is one row of the list query with its display join.
type listRow struct {
	playlist  Playlist
	songCount int
	starred   bool
	rating    *float64
}

// GetByID loads one playlist by id, parsing its rules. Rules that fail to
// parse degrade to nil (treated as unconstrained) — stored rules were
// validated at write time, so a parse failure means hand-edited data.
func GetByID(ctx context.Context, q auth.Queries, id string) (*Playlist, error) {
	var p Playlist
	var description, shareToken, rulesJSON sql.NullString
	var isSmart int
	err := q.QueryRowContext(ctx, `
		SELECT p.id, p.name, p.description, p.owner_id, u.username, p.visibility,
		       p.share_token, p.is_smart, p.rules_json, p.resolve_mode, p.created_at, p.updated_at
		FROM playlists p JOIN users u ON u.id = p.owner_id
		WHERE p.id = ?`, id).
		Scan(&p.ID, &p.Name, &description, &p.OwnerID, &p.OwnerUsername,
			&p.Visibility, &shareToken, &isSmart, &rulesJSON, &p.ResolveMode,
			&p.CreatedAt, &p.UpdatedAt)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, ErrNotFound
	}
	if err != nil {
		return nil, fmt.Errorf("load playlist: %w", err)
	}
	if description.Valid {
		p.Description = description.String
	}
	if shareToken.Valid {
		p.ShareToken = shareToken.String
	}
	p.IsSmart = isSmart == 1
	p.ResolveMode = NormalizeResolveMode(p.ResolveMode)
	if rulesJSON.Valid && rulesJSON.String != "" {
		if rules, err := ParseRules([]byte(rulesJSON.String)); err == nil {
			p.Rules = rules
		}
	}
	return &p, nil
}

// ListVisible loads the playlists a user may see in the list view: owned,
// public, and shared-with-me (v1 parity, ORDER BY updated_at DESC).
func ListVisible(ctx context.Context, q auth.Queries, userID string) ([]listRow, error) {
	rows, err := q.QueryContext(ctx, `
		SELECT p.id, p.name, p.description, p.owner_id, u.username, p.visibility,
		       p.share_token, p.is_smart, p.rules_json, p.resolve_mode, p.created_at, p.updated_at,
		       (SELECT COUNT(*) FROM playlist_songs ps WHERE ps.playlist_id = p.id) AS song_count,
		       up.starred, up.rating
		FROM playlists p
		JOIN users u ON u.id = p.owner_id
		LEFT JOIN user_playlists up ON up.user_id = ? AND up.playlist_id = p.id
		WHERE p.owner_id = ?
		   OR p.visibility = 'public'
		   OR EXISTS (SELECT 1 FROM playlist_shares ps WHERE ps.playlist_id = p.id AND ps.user_id = ?)
		ORDER BY p.updated_at DESC`, userID, userID, userID)
	if err != nil {
		return nil, fmt.Errorf("list playlists: %w", err)
	}
	defer rows.Close()
	out := []listRow{}
	for rows.Next() {
		var p Playlist
		var description, shareToken, rulesJSON sql.NullString
		var isSmart int
		var starred sql.NullInt64
		var ratingF sql.NullFloat64
		var r listRow
		if err := rows.Scan(&p.ID, &p.Name, &description, &p.OwnerID, &p.OwnerUsername,
			&p.Visibility, &shareToken, &isSmart, &rulesJSON, &p.ResolveMode,
			&p.CreatedAt, &p.UpdatedAt, &r.songCount, &starred, &ratingF); err != nil {
			return nil, fmt.Errorf("list playlists: %w", err)
		}
		if description.Valid {
			p.Description = description.String
		}
		if shareToken.Valid {
			p.ShareToken = shareToken.String
		}
		p.IsSmart = isSmart == 1
		p.ResolveMode = NormalizeResolveMode(p.ResolveMode)
		if rulesJSON.Valid && rulesJSON.String != "" {
			if rules, err := ParseRules([]byte(rulesJSON.String)); err == nil {
				p.Rules = rules
			}
		}
		r.playlist = p
		r.starred = starred.Valid && starred.Int64 == 1
		if ratingF.Valid {
			f := ratingF.Float64
			r.rating = &f
		}
		out = append(out, r)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("list playlists: %w", err)
	}
	return out, nil
}

// SongIDs returns the position-ordered member ids of a static playlist.
func SongIDs(ctx context.Context, q auth.Queries, playlistID string) ([]string, error) {
	rows, err := q.QueryContext(ctx,
		`SELECT song_id FROM playlist_songs WHERE playlist_id = ? ORDER BY position`, playlistID)
	if err != nil {
		return nil, fmt.Errorf("list playlist songs: %w", err)
	}
	defer rows.Close()
	ids := []string{}
	for rows.Next() {
		var id string
		if err := rows.Scan(&id); err != nil {
			return nil, fmt.Errorf("list playlist songs: %w", err)
		}
		ids = append(ids, id)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("list playlist songs: %w", err)
	}
	return ids, nil
}

// Create inserts a playlist and, for static playlists, its members — in ONE
// transaction (v1 B3 fix: the row insert and member rewrites must not be
// observable separately).
func Create(ctx context.Context, db *sql.DB, p *Playlist, songIDs []string) error {
	tx, err := db.BeginTx(ctx, nil)
	if err != nil {
		return err
	}
	defer tx.Rollback()
	if err := insertPlaylist(ctx, tx, p); err != nil {
		return err
	}
	if !p.IsSmart {
		if err := insertMembers(ctx, tx, p.ID, songIDs); err != nil {
			return err
		}
	}
	return tx.Commit()
}

// Update rewrites the playlist row and, when rewriteMembers is set, replaces
// the member list with exactly songIDs — in ONE transaction. Smart playlists
// always carry an empty member list; rewriteMembers clears any stale rows
// (v1 parity for smart/static conversions).
func Update(ctx context.Context, db *sql.DB, p *Playlist, songIDs []string, rewriteMembers bool) error {
	tx, err := db.BeginTx(ctx, nil)
	if err != nil {
		return err
	}
	defer tx.Rollback()
	rulesJSON := sql.NullString{}
	if p.IsSmart && p.Rules != nil {
		raw, err := json.Marshal(p.Rules)
		if err != nil {
			return fmt.Errorf("encode rules: %w", err)
		}
		rulesJSON = sql.NullString{String: string(raw), Valid: true}
	}
	isSmart := 0
	if p.IsSmart {
		isSmart = 1
	}
	shareToken := sql.NullString{}
	if p.ShareToken != "" {
		shareToken = sql.NullString{String: p.ShareToken, Valid: true}
	}
	description := sql.NullString{}
	if p.Description != "" {
		description = sql.NullString{String: p.Description, Valid: true}
	}
	res, err := tx.ExecContext(ctx, `
		UPDATE playlists
		SET name = ?, description = ?, visibility = ?, share_token = ?, is_smart = ?,
		    rules_json = ?, resolve_mode = ?, updated_at = datetime('now')
		WHERE id = ?`,
		p.Name, description, p.Visibility, shareToken, isSmart, rulesJSON,
		NormalizeResolveMode(p.ResolveMode), p.ID)
	if err != nil {
		return fmt.Errorf("update playlist: %w", err)
	}
	if n, _ := res.RowsAffected(); n == 0 {
		return ErrNotFound
	}
	if rewriteMembers {
		if _, err := tx.ExecContext(ctx,
			`DELETE FROM playlist_songs WHERE playlist_id = ?`, p.ID); err != nil {
			return fmt.Errorf("clear playlist members: %w", err)
		}
		if !p.IsSmart {
			if err := insertMembers(ctx, tx, p.ID, songIDs); err != nil {
				return err
			}
		}
	}
	return tx.Commit()
}

func insertPlaylist(ctx context.Context, tx *sql.Tx, p *Playlist) error {
	rulesJSON := sql.NullString{}
	if p.IsSmart && p.Rules != nil {
		raw, err := json.Marshal(p.Rules)
		if err != nil {
			return fmt.Errorf("encode rules: %w", err)
		}
		rulesJSON = sql.NullString{String: string(raw), Valid: true}
	}
	isSmart := 0
	if p.IsSmart {
		isSmart = 1
	}
	shareToken := sql.NullString{}
	if p.ShareToken != "" {
		shareToken = sql.NullString{String: p.ShareToken, Valid: true}
	}
	description := sql.NullString{}
	if p.Description != "" {
		description = sql.NullString{String: p.Description, Valid: true}
	}
	_, err := tx.ExecContext(ctx, `
		INSERT INTO playlists (id, name, description, owner_id, visibility, share_token, is_smart, rules_json, resolve_mode)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		p.ID, p.Name, description, p.OwnerID, p.Visibility, shareToken, isSmart,
		rulesJSON, NormalizeResolveMode(p.ResolveMode))
	if err != nil {
		return fmt.Errorf("insert playlist: %w", err)
	}
	return nil
}

// insertMembers writes the position-ordered member list. Duplicate ids are
// collapsed (a playlist is a set with positions; PRIMARY KEY would reject
// duplicates mid-transaction anyway).
func insertMembers(ctx context.Context, tx *sql.Tx, playlistID string, songIDs []string) error {
	seen := make(map[string]bool, len(songIDs))
	position := 0
	for _, id := range songIDs {
		if seen[id] {
			continue
		}
		seen[id] = true
		if _, err := tx.ExecContext(ctx,
			`INSERT INTO playlist_songs (playlist_id, song_id, position) VALUES (?, ?, ?)`,
			playlistID, id, position); err != nil {
			return fmt.Errorf("insert playlist member: %w", err)
		}
		position++
	}
	return nil
}

// Delete removes a playlist row (members and shares cascade via FK).
func Delete(ctx context.Context, q auth.Queries, playlistID string) error {
	res, err := q.ExecContext(ctx, `DELETE FROM playlists WHERE id = ?`, playlistID)
	if err != nil {
		return fmt.Errorf("delete playlist: %w", err)
	}
	if n, _ := res.RowsAffected(); n == 0 {
		return ErrNotFound
	}
	return nil
}

// ShareUpsert grants (or updates) a user share.
func ShareUpsert(ctx context.Context, q auth.Queries, playlistID, userID string, canEdit bool) error {
	edit := 0
	if canEdit {
		edit = 1
	}
	_, err := q.ExecContext(ctx, `
		INSERT INTO playlist_shares (playlist_id, user_id, can_edit) VALUES (?, ?, ?)
		ON CONFLICT(playlist_id, user_id) DO UPDATE SET can_edit = excluded.can_edit`,
		playlistID, userID, edit)
	if err != nil {
		return fmt.Errorf("share playlist: %w", err)
	}
	return nil
}

// ShareDelete revokes a user share.
func ShareDelete(ctx context.Context, q auth.Queries, playlistID, userID string) error {
	_, err := q.ExecContext(ctx,
		`DELETE FROM playlist_shares WHERE playlist_id = ? AND user_id = ?`, playlistID, userID)
	if err != nil {
		return fmt.Errorf("unshare playlist: %w", err)
	}
	return nil
}

// ShareEntries lists a playlist's shares joined with usernames (owner
// detail view).
func ShareEntries(ctx context.Context, q auth.Queries, playlistID string) ([]ShareEntry, error) {
	rows, err := q.QueryContext(ctx, `
		SELECT ps.user_id, u.username, ps.can_edit
		FROM playlist_shares ps JOIN users u ON u.id = ps.user_id
		WHERE ps.playlist_id = ?
		ORDER BY u.username`, playlistID)
	if err != nil {
		return nil, fmt.Errorf("list playlist shares: %w", err)
	}
	defer rows.Close()
	out := []ShareEntry{}
	for rows.Next() {
		var e ShareEntry
		var canEdit int
		if err := rows.Scan(&e.UserID, &e.Username, &canEdit); err != nil {
			return nil, fmt.Errorf("list playlist shares: %w", err)
		}
		e.CanEdit = canEdit == 1
		out = append(out, e)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("list playlist shares: %w", err)
	}
	return out, nil
}

// EnableShareLink moves a playlist to visibility=link with exactly token.
// Called with a freshly minted token (create + regenerate share-link).
func EnableShareLink(ctx context.Context, q auth.Queries, playlistID, token string) error {
	_, err := q.ExecContext(ctx, `
		UPDATE playlists SET visibility = 'link', share_token = ?, updated_at = datetime('now')
		WHERE id = ?`, token, playlistID)
	if err != nil {
		return fmt.Errorf("enable share link: %w", err)
	}
	return nil
}

// DisableShareLink moves a playlist back to private and clears its token:
// the ONE lifecycle rule — a token exists iff visibility == 'link'.
func DisableShareLink(ctx context.Context, q auth.Queries, playlistID string) error {
	_, err := q.ExecContext(ctx, `
		UPDATE playlists SET visibility = 'private', share_token = NULL, updated_at = datetime('now')
		WHERE id = ?`, playlistID)
	if err != nil {
		return fmt.Errorf("disable share link: %w", err)
	}
	return nil
}

// UserExists reports whether a user id exists (share target validation).
func UserExists(ctx context.Context, q auth.Queries, userID string) (bool, error) {
	var one int
	err := q.QueryRowContext(ctx, `SELECT 1 FROM users WHERE id = ?`, userID).Scan(&one)
	if errors.Is(err, sql.ErrNoRows) {
		return false, nil
	}
	if err != nil {
		return false, fmt.Errorf("user lookup: %w", err)
	}
	return true, nil
}

// HideExplicit reports the viewer's explicit-content preference; anonymous
// viewers get false (v1 parity: no preference, no filtering).
func HideExplicit(ctx context.Context, q auth.Queries, userID string) (bool, error) {
	if userID == "" {
		return false, nil
	}
	var flag int
	err := q.QueryRowContext(ctx, `SELECT hide_explicit FROM users WHERE id = ?`, userID).Scan(&flag)
	if errors.Is(err, sql.ErrNoRows) {
		return false, nil
	}
	if err != nil {
		return false, fmt.Errorf("load hide_explicit: %w", err)
	}
	return flag == 1, nil
}

// newPlaylistID mints a playlist id.
func newPlaylistID() string { return uuid.NewString() }
